contracts/src/v2/Calls.sol

```diff
@@ -0,0 +1,222 @@
+    uint8 public constant MAX_ASSETS = 8;
+
+    // Send an XCM with arbitrary assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded XCM message
+    //   * `assets` (bytes[]): Array of asset specs.
+    //
+    // Supported asset specs:
+    // * ERC20: abi.encode(0, tokenAddress, value)
+    //
+    // On Asset Hub, the assets will be received into the assets holding register.
+    //
+    // The `xcm` should contain the necessary instructions to:
+    // 1. Pay XCM execution fees, either from assets in holding,
+    //    or from the sovereign account of `msg.sender`.
+    // 2. Handle the assets in holding, either depositing them into
+    //    some account, or forwarding them to another destination.
+    //
+    function sendMessage(
+        bytes calldata xcm,
+        bytes[] calldata assets,
+        bytes calldata claimer
+    ) external {
+        _sendMessage(msg.sender, xcm, assets, claimer, msg.value);
+    }
+
+    // Register Ethereum-native token on AHP, using `xcmFeeAHP` of `msg.value`
+    // to pay for execution on AHP
+    function registerToken(address token, uint128 xcmFeeAHP) external {
+        _registerToken(token, xcmFeeAHP, 0);
+    }
+
+    // Register Ethereum-native token on AHK, using `xcmFeeAHP` and `xcmFeeAHK`
+    // of `msg.value` to pay for execution on AHP and AHK respectively.
+    function registerTokenOnKusama(address token, uint128 xcmFeeAHP, uint128 xcmFeeAHK)
+        external
+    {
+        _registerToken(token, xcmFeeAHP, xcmFeeAHK);
+    }
+
+    /*
+    * Internal functions
+    */
+
+    function _sendMessage(
+        address origin,
+        bytes memory xcm,
+        bytes[] memory assets,
+        bytes memory claimer,
+        uint256 reward
+    ) internal {
+        if (assets.length > MAX_ASSETS) {
+            revert IGateway.TooManyAssets();
+        }
+
+        bytes[] memory encodedAssets = new bytes[](assets.length);
+        for (uint256 i = 0; i < assets.length; i++) {
+            encodedAssets[i] = _handleAsset(assets[i]);
+        }
+        Ticket memory ticket = Ticket({
+            origin: origin,
+            assets: encodedAssets,
+            xcm: xcm,
+            claimer: claimer,
+            reward: reward
+        });
+        _submitOutbound(ticket);
+    }
+
+    function _registerToken(address token, uint128 xcmFeeAHP, uint128 xcmFeeAHK)
+        internal
+    {
+        // Build XCM for token registration on AHP and possibly AHK
+        // The XCM includes the necessary `PaysFee` instructions that:
+        // 1. Buy `xcmFeeAHP` worth of execution on AHP
+        // 2. Buy `xcmFeeAHK` worth of execution on AHK
+        bytes memory xcm;
+        if (xcmFeeAHK > 0) {
+            // Build XCM that is executed on AHP and forwards another message to AHK
+            xcm = bytes.concat(hex"deadbeef", abi.encodePacked(token), hex"deadbeef");
```

Comment:
`hex"deadbeef"` is this a placeholder for something?

Reply:
Will the contract build this XCM packet?

Reply:
yeah, its gonna be baked into the contract.